### ISSUE TITLE
Initialize memory with data objects or use specialized memory area

### ISSUE THREAD

**Issue by chriseth** [powdr-labs/powdr#639]
We could replace the routine that manually stores data objects in memory by a mechanism that modifies the memory machine instead: It could use a lookup into fixed data for the initial values.

Since data objects are in a read-only memory (this is a concept of the assembly language), we could even store it differently, at a place that cannot be overwritten (and thus has a simpler machine). We are in control of the actual address of the data objects, so we could use a special bit in the address to signify that location. If that bit is set, the memory machine would do a fixed lookup into that data instead of using the full memory machine mechanism.

**Comment by lvella** [comment]
I don't know how continuations are implemented, but you have to continue the proof from the memory+registers state left by the previous continuation. Whatever the mechanism used, it solves the same problem here: we have to start the computation from a predefined memory state.

**Comment by chriseth** [comment]
This is true! For continuations, though, the initial memory state would need to be a witness polynomial and not a fixed polynomial as in the case outlined above. But if we do that, we would just need to swap out the proof for the correctness of the initial memory contents.

**Comment by leonardoalt** [comment]
@lvella didn't your PR do this recently? I'll close for now with that assumption, please reopen otherwise.

**Comment by lvella** [comment]
It is not done. What we have works only with continuations' bootloader, and it is very different from OP's description. That said, I don't think such an initial memory mechanism would be much better than what we have now... maybe this is a won't fix?

**Comment by leonardoalt** [comment]
Ah right, I confused things. Yea I guess this could be an optimization if we implement things like a paper that @georgwiese mentioned recently, what was it again?

**Comment by georgwiese** [comment]
Not sure what you're talking about, perhaps this? https://eprint.iacr.org/2024/325?utm_source=substack&utm_medium=email
Victor mentioned it in the office, but I haven't read it yet.

I think it might still be a good idea to implement this as described, depending how much data it is typically and how much is actually read. With the continuations approach Lucas implemented, it still is written to memory explicitly, just as part of the bootloader and only the pages that are read in the current chunk. With the approach Chris described, this could be skipped entirely.

**Comment by leonardoalt** [comment]
Yea that one, thanks!
